python/cudf_polars/cudf_polars/dsl/expressions/base.py

```diff
@@ -87,11 +79,7 @@ def do_evaluate(
         )  # pragma: no cover; translation of unimplemented nodes trips first
 
     def evaluate(
-        self,
-        df: DataFrame,
-        *,
-        context: ExecutionContext = ExecutionContext.FRAME,
-        mapping: Mapping[Expr, Column] | None = None,
```

Comment:
https://github.com/rapidsai/cudf/pull/17941 is currently relying on the `mapping` argument to execute a "decomposed" `Expr` graph within a `Select` operation. 

Do you think the approach in 17941 makes sense, or do we want to somehow use the same decomposition logic you are using for groupby aggregations?  (You may have a general solution in mind, but this seems pretty tricky to me)

Reply:
I need to remind myself how that works. I put the mapping in as kind of a hack. Preferably one would rewrite the expression to evaluate to replace the piece with the thing that lives in the mapping

Reply:
>I need to remind myself how that works.

A `FusedExpr` basically holds a pointer to a single node in an `Expr` graph (we store this as `FusedExpr.sub_expr`). When a `FusedExpr` is evaluated, we evaluate `sub_expr`. However, as we recursively evaluate the expression graph, we must stop a soon as we get to a `FusedExpr` node.

This "stopping" mechanism is implemented with the `mapping` argument: The children of a `FusedExpr` node are the other `FusedExpr` nodes that it depends on. Since those sub-expressions will already be evaluated, we just substitute the results at evaluation time.